[Core] Implement py_executable support for runtime environments to enable better UV support (#50160)
ray-project/ray: doc/source/ray-core/handling-dependencies.rst

  ``conda`` environments must have the same Python version as the Ray cluster.  Do not list ``ray`` in the ``conda`` dependencies, as it will be automatically installed.

Using ``uv`` for package management
"""""""""""""""""""""""""""""""""""

The recommended approach for package management with `uv` in runtime environments is through `uv run`.

This method offers several key advantages:
First, it keeps dependencies synchronized between your driver and Ray workers.
Additionally, it provides full support for `pyproject.toml` including editable
packages. It also allows you to lock package versions using `uv lock`.
For more details, you can refer to the `UV scripts documentation <https://docs.astral.sh/uv/guides/scripts/>`_.

> Create a file `pyproject.toml` in your working directory like this:

Reviewer: I'm not too familiar with `uv`, but will this pattern cause the environment to be installed dynamically on worker startup? And is there any associated caching of the environment or will this happen for each newly-started worker?

I assume there is some other API that can use pre-built environments -- I would suggest we document that as the best practice for production usage.

Author: Yes it will be installed dynamically on worker startup, it is very fast and being cached (https://docs.astral.sh/uv/concepts/cache/)

```
% uv cache dir
/Users/pcmoritz/.cache/uv
```

And `uv lock` would be the best practice for freezing the environment, it will create a `uv.lock` file in the working directory that is then used on the workers too. I briefly mention it in the paragraph above but can mention it again in the example down below and tell people that this is the recommended way for locking.

Reviewer: Got it, thanks for clarifying

  ``conda`` environments must have the same Python version as the Ray cluster.  Do not list ``ray`` in the ``conda`` dependencies, as it will be automatically installed.

.. _use-uv-for-package-management:

Using ``uv`` for package management
"""""""""""""""""""""""""""""""""""

The recommended approach for package management with `uv` in runtime environments is through `uv run`.

This method offers several key advantages:
First, it keeps dependencies synchronized between your driver and Ray workers.
Additionally, it provides full support for `pyproject.toml` including editable
packages. It also allows you to lock package versions using `uv lock`.
For more details, you can refer to the `UV scripts documentation <https://docs.astral.sh/uv/guides/scripts/>`_.

Create a file `pyproject.toml` in your working directory like this:

.. code-block:: toml

  [project]

  name = "test"

  version = "0.1"

  dependencies = [
    "emoji",
    "ray",
  ]

> And then a `test.py` like this:

Reviewer 2: ```suggestion
And then a `test.py` like the following:
```

.. testcode::
  :skipif: True

  import ray

>   # We add --isolated to avoid uv problems with concurrent environment setup

Reviewer 2: ```suggestion
  # Add `--isolated` to avoid uv problems with concurrent environment setup.
```

  ray.init(runtime_env={"working_dir": ".", "py_executable": "uv run --isolated"})

  @ray.remote
  def message(entity):
      import emoji
      return emoji.emojize(entity + " rocks :thumbs_up:")

  entities = [
    "Ray",
    "Ray Serve",
    "Ray Data",
    "Ray Train",
    "Ray RLlib",
    "Ray Tune",
  ]
  results = [message.remote(entity) for entity in entities]
  for result in results:
      print(ray.get(result))

> and run the driver script with `uv run test.py`. For reproducibility it is

Reviewer 2: ```suggestion
and run the driver script with `uv run test.py`. For reproducibility it's
```

recommended to freeze the package versions by running `uv lock`.

> This workflow also supports editable packages, you can e.g. use

Reviewer 2: ```suggestion
This workflow also supports editable packages, for example, you can use
```

`uv add --editable ./path/to/package` where `./path/to/package`

> must be inside your `working_dir` so it is available to all

Reviewer 2: ```suggestion
must be inside your `working_dir` so it's available to all
```

  Note: For option (1), if the local directory contains a ``.gitignore`` file, the files and paths specified there are not uploaded to the cluster.  You can disable this by setting the environment variable `RAY_RUNTIME_ENV_IGNORE_GITIGNORE=1` on the machine doing the uploading.

- ``py_executable`` (str): Specifies the executable used for running the Ray workers. It can include arguments as well. The executable can be

>   located in the `working_dir`. This runtime environment is useful to run workers in a custom debugger or profiler as well as to run workers

Reviewer: suggest to link to the section above that shows how to use `uv`
